source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,487 @@
+                                       Runtime::Loader& runtime, Runtime::RandomGenerator& random,
+                                       const LoadBalancerSubsetInfo& subsets)
+    : lb_type_(lb_type), stats_(stats), runtime_(runtime), random_(random),
+      fallback_policy_(subsets.fallbackPolicy()), default_subset_(subsets.defaultSubset()),
+      subset_keys_(subsets.subsetKeys()), original_host_set_(host_set),
+      original_local_host_set_(local_host_set) {
+  RELEASE_ASSERT(subsets.isEnabled());
+
+  if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::DEFAULT_SUBSET &&
+      default_subset_.fields_size() == 0) {
+    // All hosts will match, so don't bother checking.
+    fallback_policy_ = envoy::api::v2::Cluster::LbSubsetConfig::ANY_ENDPOINT;
+  }
+
+  // Create filtered default subset (if necessary) and other subsets based on current hosts.
+  update(host_set.hosts(), {}, false);
+
+  // Configure future updates.
+  host_set.addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
+                                    const std::vector<HostSharedPtr>& hosts_removed) -> void {
+    update(hosts_added, hosts_removed, false);
+  });
+
+  if (local_host_set) {
+    // No need to invoke update with local = true: subset creation handles the local hosts.
+    local_host_set->addMemberUpdateCb(
+        [this](const std::vector<HostSharedPtr>& hosts_added,
+               const std::vector<HostSharedPtr>& hosts_removed) -> void {
+          update(hosts_added, hosts_removed, true);
+        });
+  }
+}
+
+HostConstSharedPtr SubsetLoadBalancer::chooseHost(const LoadBalancerContext* context) {
+  if (context) {
+    const Router::MetadataMatchCriteria* matches = context->metadataMatchCriteria();
+    if (matches) {
+      // Route has metadata_matches defined, see if we have a matching subset.
+      SubsetPtr subset = findSubset(matches->metadataMatchCriteria());
+      if (subset != nullptr && subset->lb_ != nullptr) {
+        HostConstSharedPtr host = subset->lb_->chooseHost(context);
+        // Fallback if the subset was empty (inactive), but not, for
+        // example, if all its hosts are unhealthy.
+        if (host || !subset->empty()) {
+          stats_.lb_subsets_selected_.inc();
+          return host;
+        }
+      }
+    }
+  }
+
+  if (default_host_subset_ == nullptr) {
+    return nullptr;
+  }
+
+  stats_.lb_subsets_fallback_.inc();
+  return default_host_subset_->lb_->chooseHost(context);
+}
+
+// Iterates over the given metadata matches (which must be lexically
+// sorted by key) and find a matching SubsetPtr, if any.
+SubsetLoadBalancer::SubsetPtr SubsetLoadBalancer::findSubset(
+    const std::vector<Router::MetadataMatchCriterionConstSharedPtr>& matches) {
+  LbSubsetMap* subsets = &subsets_;
+  for (size_t i = 0; i < matches.size(); i++) {
+    const Router::MetadataMatchCriterionConstSharedPtr& match = matches[i];
```

Comment:
Would it be clearer to just do a `.get()` here and work with the underlying pointer? Parsing the above type is some cognitive load.